lib/tasks/sul.rake

```diff
@@ -1,21 +1,29 @@
 
 namespace :sul do
-  desc 'rebuild authorship for selected pubs'
-  task update_contribs_in_hash: :environment do
-    #  publications = Publication.arel_table
-    #	Publication.where(publications[:pub_hash].
-    #    	matches("%okogiri%")).
-    #    	each {|publication| publication.rebuild_pub_hash }
-    Publication.find_each(&:rebuild_authorship)
-  end
-
-  desc 'rebuild authorship for selected pubs'
-  task rebuild_bad_hashes: :environment do
-    publications = Publication.arel_table
-    Publication.where(publications[:pub_hash]
-        .matches('%okogiri%'))
-      .each(&:rebuild_pub_hash)
-    # Publication.find_each {|pub| pub.rebuild_authorship }
+  desc 'Update pub_hash or authorship for all pubs'
+  # bundle exec rake sul:update_pubs['rebuild_pub_hash'] # for pub hash rebuild
+  # bundle exec rake sul:update_pubs['rebuild_authorship'] # for authorship rebuild
+  task :update_pubs, [:method] => :environment do |_t, args|
+    method = args[:method] || "rebuild_pub_hash" # default to rebuilding pub_hash, could also rebuild_authorship
+    $stdout.sync = true # flush output immediately
+    include ActionView::Helpers::NumberHelper # for nice display output and time computations in output
+    include ActionView::Helpers::DateHelper
+    total_pubs = Publication.count
+    start_time = Time.now
+    output_each = 500
+    puts "Calling #{method} for #{number_with_delimiter(total_pubs)} publications.  Started at #{start_time}.  Status update shown each #{output_each} publications."
+    Publication.find_each.with_index do |pub, index|
+      current_time = Time.now
+      elapsed_time = current_time - start_time
+      avg_time_per_pub = elapsed_time/(index+1)
+      total_time_remaining = (avg_time_per_pub * (total_pubs-index)).floor
+      if index % output_each == 0 # provide some feedback every X pubs
+        puts "...#{current_time}: on publication #{number_with_delimiter(index+1)} of #{number_with_delimiter(total_pubs)} : ~ #{distance_of_time_in_words(start_time,start_time + total_time_remaining.seconds)} left"
+      end
+      pub.send(method)
+      pub.save
+    end
+    puts "Called #{method} for #{number_with_delimiter(total_pubs)} publications.  Ended at #{Time.now}"
```

Comment:
You might need exception handling on each pub.send(method) and pub.save, so it can continue to process the next pub?  Occasional failures don't have to crash the entire batch.  But count the failures and stop if they exceed a reasonable threshold (some way to crash if everything is failing every time).